### ISSUE TITLE
how to use all_gather in training loop?

### ISSUE THREAD

**Issue by kkarrancsu** [pytorch/ignite#2504]
I have defined my train_step in the exact same way as in the [cifar10 example](https://github.com/pytorch/ignite/blob/master/examples/contrib/cifar10/main.py#L319). Is it possible to gather all of the predictions before computing the loss? I haven't seen examples of this pattern in the ignite examples (maybe I'm missing it?), but for my application, it is more optimal to compute the loss after aggregating the forward passes and targets run on multiple GPU's.  This only matters when using `DistributedDataParallel`, since `DataParallel` automatically aggregates the outputs.

**Comment by sdesrozis** [comment]
@kkarrancsu Thank you for this question. Using `idist.all_gather()` in your training loop is possible as long as the call is made by all the processes. This is a collective method call. 

I will provide an usage example asap. Sorry for the delay.

**Comment by sdesrozis** [comment]
@kkarrancsu Thanks for your answer. 

In general, `idist.all_gather()` can be used as long as the call is made collectively by all the processes. Therefore, you can use this method to gather the predictions in your training loop.

I can provide an example asap and maybe update the doc accordingly.

However, I'm not completely sure about your question. In fact, if you want to compute predictions in ddp, gather and back propagate from one proc, it won't work. You can check the internal design https://pytorch.org/docs/stable/notes/ddp.html#internal-design

**Comment by kkarrancsu** [comment]
@sdesrozis Thanks for your quick reply! Sorry if my initial question was unclear.  As an example, consider the following `DataParallel` case.

```
m = model()
m_dp = nn.DataParallel(m)
m_ddp = nn.DistributedDataParallel(m)

x = input # [batch_size, ...]
y_dp = m_dp(x)  # [batch_size, ...]
y_ddp = m_ddp(x) # [batch_size/ngpu, ...]
```
I'd like to gather all the `y_ddp` from all gpu's before computing a loss.  I hope that makes the question clear?

**Comment by kkarrancsu** [comment]
I'd like to call `loss.backward()`

**Comment by sdesrozis** [comment]
Ok so I think it won't work even if you gather the predictions. The gathering operation is not an autodiff function so it will cut the graph computation. The forward pass creates some internal states that won't be gathered too. 

Although I'm pretty sure that is answered in the PyTorch forum. Maybe I'm wrong though and I would be interested by a few discussions about this topic.

**Comment by kkarrancsu** [comment]
@sdesrozis Thanks - I will investigate based on your link and report back.

**Comment by sdesrozis** [comment]
> @sdesrozis Thanks - I will investigate based on your link and report back.

Good ! Although I’m doubtful about the link… Interesting by your feedback.

**Comment by vfdev-5** [comment]
@kkarrancsu can you provide a bit more details on what exactly you would like to do ?
In DDP, data is distributed to N processes and model is cloned. When we do the forward pass each process obtains predictions `y_preds = m_ddp(x)` on its data chunk and using a loss function and `loss.backward()` we can compute gradients that are finally sum up and applied to the model internally by pytorch DDP model wrapper.

**Comment by kkarrancsu** [comment]
Hi @vfdev-5, sure.

We are using the [Supervised Contrastive loss](https://arxiv.org/pdf/2004.11362.pdf) to train an embedding. In Eq. 2 of the paper, we see that the loss depends on the number of samples used to compute it (positive and negative). 

My colleague suggested to me that it is more optimal to compute the loss considering all examples (the entire batch), rather than considering `batch/ngpu` samples (which is what would happen when using DDP and computing loss locally to each GPU).  This is because the denominator in SupConLoss is computing the loss of negative samples, and by first aggregating all of the negative samples across gpus, you would get a more accurate loss.

**Comment by sdesrozis** [comment]
Ok I understand. You should have a look to a distributed implementation of SimCLR. See for instance 

https://github.com/Spijkervet/SimCLR/blob/cd85c4366d2e6ac1b0a16798b76ac0a2c8a94e58/simclr/modules/nt_xent.py#L7

This might give you some inspiration.

**Comment by lxysl** [comment]
> Ok I understand. You should have a look to a distributed implementation of SimCLR. See for instance
> 
> https://github.com/Spijkervet/SimCLR/blob/cd85c4366d2e6ac1b0a16798b76ac0a2c8a94e58/simclr/modules/nt_xent.py#L7
> 
> This might give you some inspiration.

This code is not so correct. Please check this issue: https://github.com/Spijkervet/SimCLR/issues/30 and my pr: https://github.com/Spijkervet/SimCLR/pull/46.
